### DISCUSSION TITLE
Query Modifier in v4

### DISCUSSION THREAD

**Discussion by MarkusLanger** [microsoft-search/pnp-modern-search#3013] [Ideas]
In version 3 there is the possibility to develop ["query modifiers"](https://microsoft-search.github.io/pnp-modern-search/v3/search-extensibility-library/getting-started/#configure-query-modifier).
I have no longer found this option for version 4. 
What possibilities are there in version 4 to change the query between the search box and the result web part?

I would like to implement something similar to the "Synonyms" in SharePoint Classic.

Procedure:
- The user enters his search term
- My code checks a list for possible synonyms
- These synonyms are included in the query.
- Query is passed to the Result Webpart
- User gets hits for his search term and the synonyms

**Comment by wobba** [comment]
You should be able to use the same query modifier project as for v3. You can also see the SPFx docs at https://docs.microsoft.com/en-us/sharepoint/dev/spfx/extensions/get-started/query-suggestion-tutorial-building. It's not web part dependent - as it's a SPFx feature.

Also see https://microsoft-search.github.io/pnp-modern-search/extensibility/custom_suggestions_provider/ for v4 info on suggestion provider for the PnP search box.

That said, we might bringer over the synonym feature from v3 to v4 as well.

**Comment by mahmoudhhassan** [comment]
Hi @wobba,

I think what we are missing in v4 is to create the query modifier for the modern search webpart (That run in the context of the webpart) not the SPFx one. 
This was possible in v3 as here (https://microsoft-search.github.io/pnp-modern-search/v3/search-extensibility-library/getting-started/#create-a-custom-query-modifier) but now this has been removed in v4 any reason why?

**Comment by wobba** [comment]
@mahmoudhhassan If you create a query modifier as per SPFx docs, can't you pick it as the source for the result web part?

Or are you thinking about a modifier for the PnP search box, not the OOB search box?

**Comment by mahmoudhhassan** [comment]
@wobba Actually, I was more thinking about a modifier for **the search results WP exactly as v3**. The SPFx works only with the out of the box search box and you need to have a custom search page for the site as per the documentation https://docs.microsoft.com/en-us/sharepoint/dev/spfx/building-search-extensions

My business scenario is very interesting, let me explain. Currently we are using the MSW to implement a lot of search driven UX components in my project. The exprince is very personalized using **"user query variables/User profile"** currently we are thinking to implement an **impersonation feature** to make it easy for content authors to know how things should look like for other users and I thought to implement **query modifier to intersect the query and replace the query variables with static values to check the behavior for other users.** I am using v3 so I don't have any issue but as my plan to upgrade to v4 soon I think this is a killer feature to bring back!

**Comment by Laetterman** [comment]
Hi all

I have exactly the same topic as @MarkusLanger: i want to modify the query (the same procedure as Markus to implement synonyms) from the PnP search box to the PnP results webpart. In v3 it was part of the extensibility function (not the query modifier!), but it is not available anymore in v4 ...

How can this be achieved in v4?

Thanks for your help
Marc

**Comment by wobba** [comment]
It can be achieved in v4 if someone contributes the functionality :)

**Comment by Laetterman** [comment]
Hi
I'd be glad to contribute but I an not that deep in development :-). We will analyze internally with a developer to see what this might mean and to get some know-how... :-)
